R/expand.R

```diff
@@ -1,78 +1,139 @@
+#' approximate dates (annotated with ? or ~).
+#' Uncertain dates may refer to several possible dates.
 #' The function "opens" these values to include all the possible dates
-#' contained in uncertain dates.
+#' implied by uncertain dates.
+#' Imprecise dates (dates only containing information on year and/or month)
+#' are also expanded to include possible dates in that year and/or month.
 #' @param x A `messydt` object.
 #' @return A list of dates, including all dates in each range or set.
 #' @export
-expand <- function(x) UseMethod("expand")
+expand <- function(x, approx_range) UseMethod("expand")
 
 #' @describeIn expand Expanding messydates
-#' @importFrom stringr str_replace_all str_split
+#' @importFrom stringr str_replace_all str_split str_detect
+#' str_extract str_remove_all
+#' @param approx_range range to expand approximate dates to, by default 3.
+#' That is, 3 days for day approximation, 3 months for month approximation, and
+#' 3 years for year approximation.
 #' @examples
-#' d <- as_messydate(c("2001-01-01", "2001-01", "2001", "2001-01-01..2001-02-02",
-#'         "{2001-01-01,2001-02-02}", "{2001-01,2001-02-02}"))
+#' d <- as_messydate(c("2008-03-25", "2012-02-27", "2001-01?", "2001~",
+#' "2001-01-01..2001-02-02", "{2001-01-01,2001-02-02}",
+#' "{2001-01,2001-02-02}", "2008-XX-31", "..2002-02-03", "2001-01-03.."))
 #' expand(d)
 #' @export
-expand.messydt <- function(x) {
-
-  x <- remove_qualifiers(x)
-
+expand.messydt <- function(x, approx_range = 3) {
+  x <- stringr::str_remove_all(x, "[:space:]")
+  x <- stringr::str_remove_all(x, "\\{")
+  x <- stringr::str_remove_all(x, "\\}")
+  x <- stringr::str_remove_all(x, "\\%")
+  x <- stringr::str_remove_all(x, "\\?") # Dubious dates are not expanded...
+  x <- expand_approximate(x, approx_range)
   x <- expand_unspecified(x)
   x <- expand_sets(x)
   x <- expand_ranges(x)
-
   x
 }
 
-remove_qualifiers <- function(dates) {
-  dates <- stringr::str_remove_all(dates, "\\?")
-  dates <- stringr::str_remove_all(dates, "\\~")
-  dates <- stringr::str_remove_all(dates, "\\%")
-  dates <- stringr::str_remove_all(dates, "[:space:]")
-  dates <- stringr::str_remove_all(dates, "\\{")
-  dates <- stringr::str_remove_all(dates, "\\}")
+expand_approximate <- function(dates, approx_range) {
+
+  dates <- ifelse(stringr::str_detect(dates, "^[:digit:]{4}\\~$"),
+                  paste0(dates, "-01-01"), dates)
+  dates <- ifelse(stringr::str_detect(dates, "^[:digit:]{4}-[:digit:]{2}\\~$"),
+                  paste0(dates, "-01"), dates)
+  dates <- ifelse(stringr::str_detect(dates,
+  "^[:digit:]{4}\\~-[:digit:]{2}-[:digit:]{2}$|
+  |^\\~[:digit:]{4}\\-[:digit:]{2}-[:digit:]{2}$"),
+                  paste0(as.Date(gsub("\\~", "", dates)) -
+                           (365 * approx_range), "..",
+                         as.Date(gsub("\\~", "", dates)) +
+                           (365 * approx_range)), dates)
+  dates <- ifelse(stringr::str_detect(dates,
```

Comment:
Please be careful when splitting regex codes in different lines, I almost certain here it will not work properly...

I can see you ran a lintr test and was trying to cut down on lines that are to big, great job! However, try to always split code lines after a comma. There are times that it is not really possible to split code lines without interfering with the results (e.g. regex lines without "|" as here), so you can leave these lines as is even if they are a bit long. Thank you!